cpp/src/lists/apply_boolean_mask.cu

```diff
@@ -0,0 +1,140 @@
+      bitmask{lcv.null_mask()}
+  {
+  }
+
+  size_type __device__ operator()(size_type i) const
+  {
+    return bit_value_or(bitmask, i, true) ? (offsets[i + 1] - offsets[i]) : 0;
+  }
+
+ private:
+  size_type num_rows;
+  offset_type const* offsets;
+  bitmask_type const* bitmask;
+};
+
+void assert_same_list_sizes(lists_column_view const& input,
+                            lists_column_view const& boolean_mask,
+                            rmm::cuda_stream_view stream)
+{
+  auto const begin = cudf::detail::make_counting_transform_iterator(
+    0,
+    [get_list_size = get_list_size{input}, get_mask_size = get_list_size{boolean_mask}] __device__(
+      size_type i) -> size_type { return get_list_size(i) != get_mask_size(i); });
+
+  CUDF_EXPECTS(thrust::reduce(rmm::exec_policy(stream), begin, begin + input.size()) == 0,
+               "Each list row must match the corresponding boolean mask row in size.");
+}
+}  // namespace
+
+std::unique_ptr<column> apply_boolean_mask(lists_column_view const& input,
+                                           lists_column_view const& boolean_mask,
+                                           rmm::cuda_stream_view stream,
+                                           rmm::mr::device_memory_resource* mr)
+{
+  CUDF_EXPECTS(input.size() == boolean_mask.size(),
+               "Boolean masks column must have same number of rows as input.");
+
+  auto const num_rows = input.size();
+
+  if (num_rows == 0) { return cudf::empty_like(input.parent()); }
+  // Note: This assert guarantees that no elements are gathered
+  // from nominally NULL input list rows.
+  assert_same_list_sizes(input, boolean_mask, stream);
+
+  auto constexpr offset_data_type = data_type{type_id::INT32};
+
+  auto filtered_child = [&] {
+    std::unique_ptr<cudf::table> tbl =
+      cudf::detail::apply_boolean_mask(cudf::table_view{{input.get_sliced_child(stream)}},
+                                       boolean_mask.get_sliced_child(stream),
+                                       stream,
+                                       mr);
+    std::vector<std::unique_ptr<cudf::column>> columns = tbl->release();
+    return std::move(columns.front());
+  };
+
+  auto output_offsets = [&] {
+    auto boolean_mask_sliced_offsets =
+      cudf::detail::slice(
+        boolean_mask.offsets(), {boolean_mask.offset(), boolean_mask.size() + 1}, stream)
+        .front();
+
+    auto const sizes         = cudf::reduction::segmented_sum(boolean_mask.get_sliced_child(stream),
+                                                      boolean_mask_sliced_offsets,
+                                                      offset_data_type,
+                                                      null_policy::EXCLUDE,
+                                                      stream);
+    auto const scalar_0      = cudf::numeric_scalar<offset_type>{0, true, stream};
+    auto const no_null_sizes = cudf::detail::replace_nulls(*sizes, scalar_0, stream);
```

Comment:
I suspect that all the size computations here can be combined into just one kernel.

Reply:
Thank you for pointing this out. I've used a `null_replacement_iterator` instead of manifesting `no_null_sizes`.